sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -574,7 +574,12 @@ object ScalaReflection extends ScalaReflection {
               "cannot be used as field name\n" + walkedTypePath.mkString("\n"))
           }
 
-          val fieldValue = Invoke(inputObject, fieldName, dataTypeFor(fieldType))
+          val fieldValue = if (!fieldType.typeSymbol.asClass.isPrimitive) {
+            Invoke(inputObject, fieldName, dataTypeFor(fieldType))
+          } else {
+            AssertNotNull(Invoke(inputObject, fieldName, dataTypeFor(fieldType)),
```

Comment:
maybe we can just add a `returnPrimitiveType` parameter in `Invoke` and update its `nullable`. The `AssertNotNull` here will never get evaluated and is only used to correct the `nullable`.